neqo-http3/src/stream_type_reader.rs

```diff
@@ -6,57 +6,120 @@
+use crate::{
+    AppError, Error, Http3StreamType, HttpRecvStream, NewStreamType, ReceiveOutput, RecvStream,
+    Res, ResetType,
+};
+use neqo_common::{qtrace, Decoder, IncrementalDecoderUint};
 use neqo_transport::Connection;
 
 #[derive(Debug)]
 pub enum NewStreamTypeReader {
-    Read {
+    ReadType {
+        push_stream_allowed: bool,
+        reader: IncrementalDecoderUint,
+        stream_id: u64,
+    },
+    ReadSessionId {
         reader: IncrementalDecoderUint,
         stream_id: u64,
     },
     Done,
 }
 
 impl NewStreamTypeReader {
-    pub fn new(stream_id: u64) -> Self {
-        NewStreamTypeReader::Read {
+    pub fn new(stream_id: u64, push_stream_allowed: bool) -> Self {
+        NewStreamTypeReader::ReadType {
+            push_stream_allowed,
             reader: IncrementalDecoderUint::default(),
             stream_id,
         }
     }
 
-    pub fn get_type(&mut self, conn: &mut Connection) -> Option<Http3StreamType> {
-        // On any error we will only close this stream!
-        loop {
-            if let NewStreamTypeReader::Read {
-                ref mut reader,
-                stream_id,
-            } = self
-            {
+    fn read(&mut self, conn: &mut Connection) -> Res<(Option<u64>, bool)> {
+        match self {
+            NewStreamTypeReader::ReadType {
+                reader, stream_id, ..
+            }
+            | NewStreamTypeReader::ReadSessionId { reader, stream_id } => loop {
                 let to_read = reader.min_remaining();
                 let mut buf = vec![0; to_read];
-                match conn.stream_recv(*stream_id, &mut buf[..]) {
-                    Ok((0, false)) => {
-                        return None;
-                    }
-                    Ok((amount, false)) => {
-                        if let Some(res) = reader.consume(&mut Decoder::from(&buf[..amount])) {
-                            *self = NewStreamTypeReader::Done;
-                            return Some(res.into());
+                match conn.stream_recv(*stream_id, &mut buf[..])? {
+                    (0, f) => break Ok((None, f)),
+                    (amount, f) => {
+                        let res = reader.consume(&mut Decoder::from(&buf[..amount]));
+                        if res.is_some() || f {
+                            break Ok((res, f));
                         }
                     }
-                    Ok((_, true)) | Err(_) => {
+                }
+            },
+            _ => Ok((None, false)),
+        }
+    }
+
+    pub fn get_type(&mut self, conn: &mut Connection) -> Res<Option<NewStreamType>> {
+        loop {
+            let (output, fin) = self.read(conn)?;
+            if output.is_none() {
+                if fin {
+                    *self = NewStreamTypeReader::Done;
+                    break Err(Error::HttpStreamCreation);
+                }
+                break Ok(None);
+            }
+            let output = output.unwrap();
```

Comment:
I think that we can do:

```rust
let output = if let Some(o) = output {
    o
} else {
    if fin { ... }
    break Ok(None);
};
```
Also, if breaking out of a loop produces a return value, I would prefer that you use "return" rather than "break" as it makes it easier to understand that this is the end.  With "break" you need to check that the loop is the only statement in the function, "return" is unambiguous.